src/cpp/session/modules/jobs/AsyncRJobManager.cpp

```diff
@@ -0,0 +1,149 @@
+/*
+ * AsyncRJobManager.cpp
+ *
+ * Copyright (C) 2009-19 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include "AsyncRJobManager.hpp"
+
+#include <r/RExec.hpp>
+
+#include <session/jobs/JobsApi.hpp>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules { 
+namespace jobs {
+namespace {
+
+// Vector of currently active async R jobs
+std::vector<boost::shared_ptr<AsyncRJob> > s_jobs;
+
+} // anonymous namespace
+
+
+AsyncRJob::AsyncRJob(const std::string& name):
+   completed_(false),
+   cancelled_(false),
+   name_(name)
+{
+}
+
+void AsyncRJob::registerJob()
+{
+   Error error;
+   r::sexp::Protect protect;
+
+   // create the actions list
+   SEXP actions = R_NilValue;
+   error = r::exec::RFunction(".rs.scriptActions").call(&actions, &protect);
+   if (error)
+      LOG_ERROR(error);
+
+   // add the job -- currently idle until we get some content from it
+   job_ = addJob(name_, "", "", 0, JobIdle, JobTypeSession, false, actions, true, {});
+}
+
+void AsyncRJob::onStderr(const std::string& output)
+{
+   if (job_)
+      job_->addOutput(output, true);
+}
+
+void AsyncRJob::onStdout(const std::string& output)
+{
+   if (job_)
+      job_->addOutput(output, false);
+}
+
+void AsyncRJob::onCompleted(int exitStatus)
+{
+   // if the job has not yet been marked complete, do so now
+   if (!job_->complete())
+   {
+      setJobState(job_, exitStatus == 0 ?  
+         JobState::JobSucceeded : JobState::JobFailed);
```

Comment:
When would the job not be in a completed state here? Do we need to check if it was cancelled?

Reply:
This is fallback behavior so that classes that derive from AsyncRJob get reasonable defaults out of the box. In practice the two actual derived classes both do something more nuanced.

I do think it'd make sense to use cancelled here if known! I'll make that change.

Reply:
LGTM!